### ISSUE TITLE
Apply FREI-PI

### ISSUE THREAD

**Issue by smol-ninja** [sablier-labs/flow#146]
Find "one" core invariant that can be verified at the end of every execution.

### Reading
1. https://www.nascent.xyz/idea/youre-writing-require-statements-wrong for ideas
2. https://github.com/orgs/sablier-labs/discussions/7

**Comment by smol-ninja** [comment]
The below _amount check_ must always hold true at the end of every function.

##### For a given `streamId`:

$\text{stream balance} = \text{total deposited} - \text{total withdrawn} - \text{total refunded}$

If there is an attack that can leak funds through one of the functions, the execution would update either one of the above values and the invariant would fail.

While stream balance is at our disposal, we would have to store the remaining three in storage and so an increase in gas cost (5k per variable => 15k in total).

Thoughts on this?

**Comment by smol-ninja** [comment]
Yes you are right. Each of `deposit`, `refund` and `withdraw` will see an increase of 9300 (= 1 SSTORE + 3 cold SLOAD + 1 hot SLOAD) whereas other functions will see an increase of 8400 ( = 4 cold SLOAD). No function will write to more than one variable at a time.

Defining these variables would also mean removing `aggregateBalance` introduced by https://github.com/sablier-labs/flow/pull/234. `aggregateBalance` was chosen to replace having 3 variables as discussed in https://github.com/sablier-labs/company-discussions/discussions/11#discussioncomment-8243522.

**Comment by PaulRBerg** [comment]
Great, thanks for sharing the quantitative data.

That ~9k gas is an acceptable cost for building an 'invariant wall' around the protocol.

Are your gas estimates inclusive of the removal of `aggregateBalance`? Presumably, removing that variable would lead to a decrease in gas costs.

**Comment by smol-ninja** [comment]
> Are your gas estimates inclusive of the removal of aggregateBalance? Presumably, removing that variable would lead to a decrease in gas costs

No it does not. Plus there will be some arithmetic gas cost as well. But in the end, it seems to be well below 10k.

**Comment by PaulRBerg** [comment]
Great, well in this case, I'm in favor of using FREI-PI.

**Comment by smol-ninja** [comment]
Sounds good. Should we add it in v1.1.0 instead of v1.0.0?

**Comment by PaulRBerg** [comment]
Yep

**Comment by andreivladbrg** [comment]
> I think we should not implement FREI-PI in our protocol since it's not feasible. It's limited by our design, where one contract manages all the streams.

as discussed on slack, i agree with this conclusion.

also, we have an [inv test for it](https://github.com/sablier-labs/flow/blob/1917f51701c440ca943793b63acd44f36fcfa54d/tests/invariant/Flow.t.sol#L201), which technically means it should be caught during tests

**Comment by PaulRBerg** [comment]
Excellent observation with the overflow risk, @smol-ninja.

> we have an [inv test for it](https://github.com/sablier-labs/flow/blob/1917f51701c440ca943793b63acd44f36fcfa54d/tests/invariant/Flow.t.sol#L201), which technically means it should be caught during tests

An onchain invariant is a stronger and harder guarantee compared to an offchain one. That was the very essence of FREI-PI, but alas it doesn't work in this case.
